core/domain/stats_services.py

```diff
@@ -22,52 +22,72 @@
+    stats.
+
+    Args:
+        exp_id: str. ID of the exploration.
+        exp_version: int. Version of the exploration.
+        aggregated_stats: dict. Dict representing an ExplorationStatsModel
+            instance with stats aggregated in the frontend.
     """
     exploration_stats = get_exploration_stats_by_id(
         exp_id, exp_version)
 
-    if event_type == feconf.EVENT_TYPE_START_EXPLORATION:
-        exploration_stats.num_starts_v2 += 1
-    elif event_type == feconf.EVENT_TYPE_ACTUAL_START_EXPLORATION:
-        exploration_stats.num_actual_starts_v2 += 1
-    elif event_type == feconf.EVENT_TYPE_COMPLETE_EXPLORATION:
-        exploration_stats.num_completions_v2 += 1
-    elif event_type == feconf.EVENT_TYPE_ANSWER_SUBMITTED:
+    exploration_stats.num_starts_v2 += aggregated_stats['num_starts']
+    exploration_stats.num_completions_v2 += aggregated_stats['num_completions']
+    exploration_stats.num_actual_starts_v2 += aggregated_stats[
+        'num_actual_starts']
+
+    for state_name in aggregated_stats['state_stats_mapping']:
+        exploration_stats.state_stats_mapping[
+            state_name].total_answers_count_v2 += aggregated_stats[
+                'state_stats_mapping'][state_name]['total_answers_count']
+        exploration_stats.state_stats_mapping[
+            state_name].useful_feedback_count_v2 += aggregated_stats[
+                'state_stats_mapping'][state_name]['useful_feedback_count']
         exploration_stats.state_stats_mapping[
-            state_name].total_answers_count_v2 += 1
-        if update_params['feedback_is_useful']:
-            exploration_stats.state_stats_mapping[
-                state_name].useful_feedback_count_v2 += 1
-    elif event_type == feconf.EVENT_TYPE_STATE_HIT:
+            state_name].total_hit_count_v2 += aggregated_stats[
+                'state_stats_mapping'][state_name]['total_hit_count']
         exploration_stats.state_stats_mapping[
-            state_name].total_hit_count_v2 += 1
-        if update_params['is_first_hit']:
-            exploration_stats.state_stats_mapping[
-                state_name].first_hit_count_v2 += 1
-    elif event_type == feconf.EVENT_TYPE_STATE_COMPLETED:
+            state_name].first_hit_count_v2 += aggregated_stats[
+                'state_stats_mapping'][state_name]['first_hit_count']
         exploration_stats.state_stats_mapping[
-            state_name].num_completions_v2 += 1
-    elif event_type == feconf.EVENT_TYPE_SOLUTION_HIT:
+            state_name].num_times_solution_viewed_v2 += aggregated_stats[
+                'state_stats_mapping'][state_name]['num_times_solution_viewed']
         exploration_stats.state_stats_mapping[
-            state_name].num_times_solution_viewed_v2 += 1
+            state_name].num_completions_v2 += aggregated_stats[
+                'state_stats_mapping'][state_name]['num_completions']
 
-    save_stats_model(exploration_stats)
+    transaction_services.run_in_transaction(
```

Comment:
Is this the only function that uses save_stats_model? If so let's rename state_stats_model to save_stats_model_transactional and just call that directly. Inside that function there should be an inner function which gets run in a transaction. There are examples of this type of pattern in core/jobs.py.

This ensures that the function in question will always be run transactionally.

Reply:
Done.